### ISSUE TITLE
Set Default values and table names with DataAnnotations

### ISSUE THREAD

**Issue by DiomedesDominguez** [dotnet/efcore#19546]
### Is your feature request related to a problem? 
I was trying to keep DataAnnotations from a legacy code we have made with .NET Framework, but rewriting all the defaults constraints for MS SQL as fluent in the OnModelCreating method was awful.

### Describe the solution you'd like
Something simple, like modelbuilder.SetDefaults() and read the table attribute and the attributes of their properties.

### Additional context
I made an extension method that works well, including the Identity tables:
https://gist.github.com/DiomedesDominguez/797711e093fd15eff4265ae9d5dc27b9

**Comment by ajcvickers** [comment]
@DiomedesDominguez Can you provide some more details on which data annotations are behaving differently in .NET Core?

**Comment by DiomedesDominguez** [comment]
@ajcvickers in the [documentation](https://docs.microsoft.com/en-us/ef/core/modeling/generated-properties?tabs=data-annotations#value-generated-on-add-1) we read that [DatabaseGenerated(DatabaseGeneratedOption.Identity)] for a DateTime property does not guarantee that EF will setup the actual mechanism to generate values, so we must rely on Fluent API.

Right now, in .NET Framework, Fluent API is an option, not a must, so I propose to keep the similarity between the two by allowing to those that have .NET Framework code to be able to migrate to .NET Core effortless.

**Comment by ajcvickers** [comment]
@DiomedesDominguez I'll investigate the differences between EF6 and EF Core here. However, a few things to note:
* `[DatabaseGenerated(DatabaseGeneratedOption.Identity)]` should definitely work for GUIDs on key properties
* I'm pretty sure that `DefaultValue` was ignored by EF6
* The configuration possible using data annotations has always been a subset of the configuration possible using the model building API. EF 4.1 was released this way in 2011, and every version since then has followed the same approach, including EF Core. However, there are some differences in what can be configured with data annotations between EF6 and EF Core. See #10864 for more details.

**Comment by AraHaan** [comment]
Honestly I think it should be possible for the scaffolder (and EFCore itself at runtime) to use ``System.ComponentModel.DefaultValueAttribute`` when it's configured to use data annotations over it generating an call to ``.HasDefaultValueSql`` (have it call it internally at runtime when it sees a member marked with it for us).

**Comment by DiomedesDominguez** [comment]
> Honestly I think it should be possible for the scaffolder (and EFCore itself at runtime) to use `System.ComponentModel.DefaultValueAttribute` when it's configured to use data annotations over it generating an call to `.HasDefaultValueSql` (have it call it internally at runtime when it sees a member marked with it for us).

Exactly my point: in EF6 can be used in conjunction with ```[DatabaseGenerated(DatabaseGeneratedOption.Identity)]  ``` but in EFCore not.

**Comment by AraHaan** [comment]
Yep, In my project I invented my own ``DefaultValue`` attribute:

<details>
  <summary>Microsoft.EntityFrameworkCore.DefaultValueSqlAttribute:</summary>

```cs
namespace Microsoft.EntityFrameworkCore;

using System.Reflection;

[AttributeUsage(AttributeTargets.Property)]
internal class DefaultValueSqlAttribute : Attribute
{
    public DefaultValueSqlAttribute(string defaultValueSql)
    {
        this.DefaultValueSql = defaultValueSql;
    }

    public string DefaultValueSql { get; }

    internal static void SetWithModelBuilder(Type entityType, ModelBuilder modelBuilder)
    {
        var prop = entityType.GetProperties(
            BindingFlags.Instance);
        foreach (var property in prop)
        {
            var attribute = property.GetCustomAttribute<DefaultValueSqlAttribute>();
            if (attribute is not null)
            {
                _ = modelBuilder.Entity(
                    entityType,
                    entity =>
                    {
                        entity.Property(property.Name).HasDefaultValueSql(attribute.DefaultValueSql);
                    });
            }
        }
    }
}
```
</details>

My own DateTimeKind attribute as well:
<details>
  <summary>Microsoft.EntityFrameworkCore.DateTimeKindAttribute:</summary>

```cs
namespace Microsoft.EntityFrameworkCore;

using System.Reflection;

[AttributeUsage(AttributeTargets.Property)]
internal class DateTimeKindAttribute : Attribute
{
    public DateTimeKindAttribute()
        : this(DateTimeKind.Unspecified)
    {
    }

    public DateTimeKindAttribute(DateTimeKind dateTimeKind)
    {
        this.DateTimeKind = dateTimeKind;
    }

    public DateTimeKind DateTimeKind { get; }

    // default this to false as DateTimes are not nullable by default.
    public bool IsNullable { get; set; } = false;

    internal static void SetWithModelBuilder(Type entityType, ModelBuilder modelBuilder)
    {
        var prop = entityType.GetProperties(
            BindingFlags.Instance);
        foreach (var property in prop)
        {
            var attribute = property.GetCustomAttribute<DateTimeKindAttribute>();
            if (attribute is not null)
            {
                _ = modelBuilder.Entity(
                    entityType,
                    entity =>
                    {
                        if (attribute.DateTimeKind.Equals(DateTimeKind.Utc))
                        {
                            // property uses UTC Time only.
                            entity.Property(property.Name).UsesUtcTime(
                                includeNullable: attribute.IsNullable,
                                includeNonNullable: !attribute.IsNullable);
                        }
                        else if (attribute.DateTimeKind.Equals(DateTimeKind.Local))
                        {
                            // property uses Local Time only.
                            entity.Property(property.Name).UsesLocalTime(
                                includeNullable: attribute.IsNullable,
                                includeNonNullable: !attribute.IsNullable);
                        }
                    });
            }
        }
    }
}
```
</details>

Then in my DbContext OnModelCreatingPartial method I then simply do this:
```cs
    partial void OnModelCreatingPartial(ModelBuilder modelBuilder)
    {
        DefaultValueSqlAttribute.SetWithModelBuilder(typeof(Role), modelBuilder);
        DateTimeKindAttribute.SetWithModelBuilder(typeof(ExpiringVerification), modelBuilder);
        DateTimeKindAttribute.SetWithModelBuilder(typeof(Patron), modelBuilder);
    }
```

And then I updated my t4 (text) templates to add in ``DateTimeKind`` when ``property.ClrType`` is ``System.DateTime`` and then I check if it's nullable to include ``, IsNullable = true`` when it is nullable on the generated entity property.

I will need to determine how I can go about having it dump the default value sql for my new attribute though from my text templates to avoid needing to do the manual call to ``HasDefaultValueSql`` and instead have it called from ``SetWithModelBuilder`` from within my new attribute instead.

Hopefully I could PR in these 2 attributes into EFCore for people to use (and make SetWithModelBuilder in them public for now). Of course in order to use them you must provide your own text templates (or I could PR mine to make the ones I made into default ones).

**Comment by roji** [comment]
@DiomedesDominguez 

>> Honestly I think it should be possible for the scaffolder (and EFCore itself at runtime) to use System.ComponentModel.DefaultValueAttribute when it's configured to use data annotations over it generating an call to .HasDefaultValueSql (have it call it internally at runtime when it sees a member marked with it for us).
>
> Exactly my point: in EF6 can be used in conjunction with [DatabaseGenerated(DatabaseGeneratedOption.Identity)] but in EFCore not.

Can you please post a minimal, runnable code sample of what isn't working for you? As indicated above by @ajcvickers, `[DatabaseGenerated(DatabaseGeneratedOption.Identity)]` works on Guid key properties, `[DefaultValue]` was not recognized by EF6 ([see docs](https://docs.microsoft.com/en-us/ef/ef6/modeling/code-first/data-annotations)), and I'm not sure what using them in conjunction would mean exactly.

@AraHaan

> Honestly I think it should be possible for the scaffolder (and EFCore itself at runtime) to use System.ComponentModel.DefaultValueAttribute when it's configured to use data annotations over it generating an call to .HasDefaultValueSql

Assuming you mean "DefaultValue" rather than "DefaultValueSql", that would require EF to identify when the database column's default SQL represents a simple value, and to parse it into a .NET literal value. For example, imagine there's a default value timestamp in a column; EF Core would need to recognize that and parse it into a .NET DateTime representation. This is both very hard to do reliably across all possible database types, and provides very little value over just embedding the SQL into DefaultValueSql.

Having DefaultValue/DefaultValueSql as attributes (rather than Fluent API) is another question.

> And then I updated my t4 (text) templates to add in DateTimeKind when property.ClrType is System.DateTime and then I check if it's nullable to include , IsNullable = true when it is nullable on the generated entity property.

It's worth mentioning that if the goal is to set the Kind on all DateTime properties, setting up a value converter via [model bulk configuration](https://docs.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration) is probably a better way to do it, which doesn't require any attributes.

**Comment by AraHaan** [comment]
> Having DefaultValue/DefaultValueSql as attributes (rather than Fluent API) is another question.
> 

Agreed, having `DefaultValueSql` as an attribute is the most logical option. For me my codebase uses ``DefaultValueSql`` only at the moment in 1 spot and that is in the attribute where the attribute is used on 1 of my entities on my DB.

> > And then I updated my t4 (text) templates to add in DateTimeKind when property.ClrType is System.DateTime and then I check if it's nullable to include , IsNullable = true when it is nullable on the generated entity property.
> 
> It's worth mentioning that if the goal is to set the Kind on all DateTime properties, setting up a value converter via [model bulk configuration](https://docs.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration) is probably a better way to do it, which doesn't require any attributes.

While that is an option, let's say for example all of my tables in my DB uses UTC, but then someone in the future adds a new table that requires ``DateTimeKind.Local`` to be used with it (and all new values inserted into the table). With the bulk config, you are effectively limiting that table also to UTC time. Also many developers might prefer UTC, while others use local time (and convert UTC to local time).

**Comment by roji** [comment]
>> Having DefaultValue/DefaultValueSql as attributes (rather than Fluent API) is another question.
>
> Agreed, having DefaultValueSql as an attribute is the most logical option. For me my codebase uses DefaultValueSql only at the moment in 1 spot and that is in the attribute where the attribute is used on 1 of my entities on my DB.

Assuming we implement attributes here, I don't think there's a reason to not have a [DefaultValue] attribute; it's just that we would **scaffold** it, just like we don't scaffold the `HasDefaultValue` Fluent API.

> With the bulk config, you are effectively limiting that table also to UTC time.

You can override the bulk config with a specific configuration for that table. It's effectively just a default for your model.

**Comment by AraHaan** [comment]
I have decided to make the following changes to the efcore codebase locally (and plan to PR them when I can add tests):

- EFCore.Abstractions:
  - DateTimeKindAttribute
  - DefaultValueSqlAttribute
- EFCore:
  - DateTimeKindAttributeConvention (implements IEntityTypeAddedConvention, IEntityTypeBaseTypeChangedConvention, IModelFinalizingConvention)
  - PropertyBuilder (added ``PropertyBuilder HasDateTimeKind(DateTimeKind kind, bool includeNullable = true, bool includeNonNullable = true);``) (public api and fluent style version of the attribute when used on an property)
  - InternalPropertyBuilder (added ``InternalPropertyBuilder? HasDateTimeKind(DateTimeKind kind, bool includeNullable = true, bool includeNonNullable = true, bool fromDataAnnotation = false);``) (internal api)
  - EntityTypeBuilder (added ``EntityTypeBuilder HasDateTimeKind(DateTimeKind kind, bool includeNullable = true, bool includeNonNullable = true);``) (public api and fluent style version of the attribute when used on the entity type)
  - InternalEntityTypeBuilder (added ``InternalEntityTypeBuilder HasDateTimeKind(DateTimeKind kind, bool includeNullable = true, bool includeNonNullable = true, bool fromDataAnnotation = false);``) (internal api)
  - IConventionPropertyBuilder, IConventionEntityTypeBuilder (added ``HasDateTimeKind`` as well which match the Internal versions of them (except return type))
  - ProviderConventionSetBuilder (required changes to add the new attributes)

Note: I have not ran my code *yet* to verify that it all works with my local changes.

**Comment by roji** [comment]
I don't see anything above that's specific to DateTimeKind in any way: setting up the same value converter on both nullable and non-nullable properties is a general concern with any value converter, so it wouldn't justify adding API that's specific to DateTimeKind. I also don't think it's appropriate to have an attribute as a default Kind setting on an entity type. Once again, managing the Kind is nothing more and nothing less than a value converter.

But note that we do intend to implement custom conventions in 7.0 (#214); that would allow you to define attributes and add any conventions you want as the application, without needing to do any changes in EF Core itself.
